src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,305 @@
+			}
+
+			bool wroteSomething = WriteChildren( @namespace );
+
+			FinishBlock( wroteSomething, closeBrace );
+
+			return wroteSomething;
+		}
+
+		private bool WriteType( TypeDeclarationSyntax decl )
+			=> WriteType( decl, NoneToken );
+
+		private bool WriteType(
+			TypeDeclarationSyntax decl,
+			SyntaxToken secondaryKeyword
+		) {
+			var preambleCountBefore = m_preambles.Count;
+
+			// This should either get used inside a call to WriteMethods either
+			// somewhere down the call to WriteChildren here or the direct call
+			// to WriteMethods here, or it will be discarded inside
+			// CloseBraceIfNecessary.
+			AddPartialTypeToPreamble(
+				decl,
+				keyword1: decl.Keyword,
+				// for "record struct" etc.
+				keyword2: secondaryKeyword,
+				identifier: decl.Identifier,
+				typeParameterList: decl.TypeParameterList,
+				constraintClauses: decl.ConstraintClauses,
+				openBrace: decl.OpenBraceToken
+			);
+
+			var wroteSomethingInChild = WriteChildren( decl );
+			var wroteMethodsForMe = WriteMethods( decl );
+
+			var wroteSomething = wroteSomethingInChild || wroteMethodsForMe;
+
+			FinishBlock( wroteSomething, decl.CloseBraceToken );
+
+			CheckPremableInvariant( wroteSomething, preambleCountBefore );
+
+			return wroteSomething;
+		}
+
+		private void AddPartialTypeToPreamble(
+			SyntaxNode decl,
+			SyntaxToken keyword1,
+			SyntaxToken keyword2,
+			SyntaxToken identifier,
+			TypeParameterListSyntax? typeParameterList,
+			SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses,
+			SyntaxToken openBrace
+		) {
+			// Create the partial keyword and grab all leading trivia from the node
+			var partialKeyword = CreatePartialKeyword( decl );
+
+			// Clear any leading trivia from the keyword, if it exists.
+			// Only the first token on the line should have leading trivia, and
+			// we've moved that onto partialKeyword.
+			// This could get ugly if someone did
+			//
+			//   sealed class
+			//      Foo {
+			keyword1 = keyword1.WithLeadingTrivia( EmptyTrivia );
+			keyword2 = keyword2.WithLeadingTrivia( EmptyTrivia );
+			identifier = identifier.WithLeadingTrivia( EmptyTrivia );
+
+			m_preambles.AddLast(
+				new PieceOfSyntax[] {
+					partialKeyword,
+					keyword1,
+					keyword2,
+					identifier,
+					(typeParameterList != null ? typeParameterList : NoneToken ),
+					openBrace
+				}
+			);
+		}
+
+		private static SyntaxToken CreatePartialKeyword( SyntaxNode decl )
+			=> SyntaxFactory.Token( SyntaxKind.PartialKeyword )
+				.WithLeadingTrivia( decl.GetLeadingTrivia() )
+				.WithTrailingTrivia( SyntaxFactory.Space );
+
+		private void FinishBlock( bool wroteSomething, SyntaxToken? closeBrace ) {
+			if( wroteSomething ) {
+				// Add the closing } if it exists
+				if( closeBrace.HasValue ) {
+					m_out.Append( closeBrace.Value.ToFullString() );
+				}
+			} else {
+				// Throw away the start of the decl that we never wrote
+				m_preambles.RemoveLast();

```

Comment:
Could consider holding `PieceOfSyntax[]` to pass it into `FinishBlock` to assert that `RemoveLast` is removing what we're expecting.

Reply:
Hm, interesting.

Reply:
I mean, as long as it's working and the tests show it is, that's also pretty okay. Just we're asserting some things already, so have it on mind etc etc

Reply:
Yeah. I like the idea/goal of making the invariants more obviously correct. I fiddled a bit now and I couldn't get it to a place where I was totally happy, but I'm also multi-tasking on a Zoom call. I'm going to keep fiddling with it and possibly follow-up.